### PR TITLE
convert route package to kotlin

### PR DESCRIPTION
**Pull request by abhishek1508**
## Description

Convert route package to kotlin (https://github.com/mapbox/mapbox-navigation-android/issues/2208)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Convert route package to kotlin

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package com.mapbox.services.android.navigation.v5.route
+
+import android.location.Location
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+
+/**
+ * This class can be subclassed to provide custom logic for checking / determining
+ * new / faster routes while navigating.
+ *
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt]: NIT not necessary

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * To provide your implementation,
+ * use [com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation.setFasterRouteEngine].
+ *
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt]: NIT Not necessary

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * [FasterRoute.shouldCheckFasterRoute] determines how quickly a
+ * new route will be fetched by [RouteFetcher].
+ *
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt]: NIT Not necessary

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * [FasterRoute.isFasterRoute] determines if the new route
+ * retrieved by [RouteFetcher] is actually faster than the current route.
+ *
+ */
+abstract class FasterRoute {
+
+    /**
+     * This method determine if a new [DirectionsResponse] should
+     * be retrieved by [RouteFetcher].
+     *
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRoute.kt]: Same as above and below

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+import com.mapbox.services.android.navigation.v5.navigation.NavigationConstants.NAVIGATION_CHECK_FASTER_ROUTE_INTERVAL
+import com.mapbox.services.android.navigation.v5.navigation.NavigationConstants.NAVIGATION_MEDIUM_ALERT_DURATION
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import java.util.Date
+import java.util.concurrent.TimeUnit
+
+class FasterRouteDetector : FasterRoute() {
+
+    companion object {
+
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt]: NIT not necessary

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            val newRoute = response.routes()[0]
+
+            if (hasLegs(newRoute)) {
+                // Extract the first leg
+                newRoute.legs()?.let { routeLegList ->
+                    val routeLeg = routeLegList[0]
+                    if (hasAtLeastTwoSteps(routeLeg)) {
+                        routeLeg.steps()?.let {
+                            // Extract the first two steps
+                            val firstStep = routeLeg.steps()!![0]
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt]: We could use `it` here or a lambda `steps ->` and remove `routeLeg.steps()!!`

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                            val secondStep = routeLeg.steps()!![1]
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt]: Same as above

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+import timber.log.Timber
+
+/**
+ * This class can be used to fetch new routes given a [Location] origin and
+ * [RouteOptions] provided by a [RouteProgress].
+ */
+class RouteFetcher {
+
+    companion object {
+
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt]: NIT not necessary

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        override fun onResponse(call: Call<DirectionsResponse>, response: Response<DirectionsResponse>) {
+            updateListeners(response.body(), routeProgress)
+        }
+
+        override fun onFailure(call: Call<DirectionsResponse>, throwable: Throwable) {
+            updateListenersWithError(throwable)
+        }
+    }
+
+    constructor(context: Context, accessToken: String) {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt]: Could we convert this constructor into a primary constructor?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+            // Check for both valid route and step durations remaining
+            return validRouteDurationRemaining(routeProgress) && validStepDurationRemaining(routeProgress)
+        }
+        return false
+    }
+
+    override fun isFasterRoute(response: DirectionsResponse, routeProgress: RouteProgress): Boolean {
+        if (validRouteResponse(response)) {
+            val currentDurationRemaining = routeProgress.durationRemaining()
+            val newRoute = response.routes()[0]
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/FasterRouteDetector.kt]:
```suggestion
            val newRoute = response.routes().firstOrNull()
```

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        return approaches
+    }
+
+    private fun invalid(context: Context?, location: Location?, routeProgress: RouteProgress?): Boolean {
+        return context == null || location == null || routeProgress == null
+    }
+
+    private fun updateListeners(response: DirectionsResponse?, routeProgress: RouteProgress?) {
+        for (listener in routeListeners) {
+            response?.let { directionsResponse ->
```
Comment (RingerJK) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteFetcher.kt]:
```suggestion
            response?.routeProgress?.let { routeProgress ->
```

**Pull request merged**